### ISSUE TITLE
Confusing behavior when using relative paths in deleted directories

### ISSUE THREAD

**Issue by Gellardo** [ajeetdsouza/zoxide#290]
# Short description
I aliased `z` over `cd` and after deleting the current directory in another terminal, I could no longer cd/z out of the directory with relative paths (prints `zoxide: no match found`). Using absolute paths instead works.

# Expected
I think it would be helpful for zoxide to realize the situation and adjust its output accordingly.
Since `z` makes jumping to other directories so easy, I do not think it is necessary to copy `cd` behavior.
Just make it clear that relative paths do not work right now/I am in a deleted directory/... and don't use the generic "not found" answer.

I know behavior in deleted directories is always a little weird.
When running into this I was very confused how the parent directory `..` could not be found.
Normal cd (zsh builtin) behavior is to just walk up the deleted paths normally (though `pwd` is wonky).
Once we are back in a normal directory, everything works fine.

# Reproduce
```
mkdir -p tmp/tmp2
cd tmp/tmp2
rm -r ../../tmp/tmp2
cd ..
# zoxide: no match found
cd ../..
# zoxide: no match found
cd ~
# (ok, no output)
cd -
# __zoxide_cd:cd:2: no such file or directory: /Users/frank/tmp/tmp2 (ok, meaningful error)
```
Only happens if at least 2 directories deep.

## Versions and Co
```
zoxide --version
zoxide v0.7.8
```
Installed via Homebrew on Mac, `eval "$(zoxide init zsh)"` and `alias cd='z'`.

**Comment by ajeetdsouza** [comment]
Hey @Gellardo, I understand your point, but I'm not sure how this would work.

`z` operates by checking if your query refers to an existing directory. If yes, it does a `cd`, if no, it does a "jump". Now if you were in `/tmp/foo` which has been deleted, and were trying to go to `/tmp/foo/bar`, you'd type in `z bar`, which would either jump to some other directory called `bar` or tell you that no match was found. How would zoxide know in this case if you wanted to make a query or cd into a relative path?

> Just make it clear that relative paths do not work right now

This isn't exactly true. If you were in `/tmp/foo` and deleted `/tmp/foo`, you'd still be able to use `z ..` to get out of it. That's a relative path. In your case, `z ../..` would have worked as well.

> I am in a deleted directory

Even if I added this warning into zoxide, it will be too late by the time you receive it -- it would warn you and then jump to another directory. In your case, if there was a directory on your system with `..` somewhere in its name, `z` would have taken you there.

> When running into this I was very confused how the parent directory .. could not be found.

I think the solution is that a deleted directory warning should be a part of your _prompt_ rather than a part of zoxide. My prompt shows me a lock icon if I'm in a read-only or deleted directory, which usually works for me.

**Comment by Gellardo** [comment]
That is acceptable, adjusting my prompt is actually a great idea.

> This isn't exactly true. If you were in /tmp/foo and deleted /tmp/foo, you'd still be able to use z .. to get out of it. That's a relative path. In your case, z ../.. would have worked as well.

Nitpick: No `z ../..` did not work. One level `z ..` in `tmp` works but if i am in `tmp/tmp2` using `z ../..` does not work.

**Comment by Gellardo** [comment]
Actually i just tried something else to see if queries work (which they don't which you probably already know):

```
mkdir tmp/tmp2
cd tmp/tmp2
rm -r ../../tmp
z <my username> # zoxide: no match found
zi #works, i can select a directory
```
Again, surprising but not unsolvable.

**Comment by ajeetdsouza** [comment]
> `z ../..` did not work.

Which shell are you on? This works for me:

```sh
mkdir -p /tmp/tmp1/tmp2
z /tmp/tmp1/tmp2

rm -rf /tmp/tmp1
z ../.. # succeeds, takes me to /tmp
```

> `z <my username> # zoxide: no match found`

This doesn't work anyway. zoxide doesn't store the home directory in the database by default, because:

- It always ends up getting a very high score and overshadowing the other directories
- It's already easily accessible via `z ~` or simply just `z`

However, other queries with `z` should work.

**Comment by Gellardo** [comment]
I'm using zsh 5.8

**Comment by ajeetdsouza** [comment]
Did these commands not work, then? I'm on zsh too and it works just fine.

> ```sh
> mkdir -p /tmp/tmp1/tmp2
> z /tmp/tmp1/tmp2
> 
> rm -rf /tmp/tmp1
> z ../.. # succeeds, takes me to /tmp
> ```

**Comment by Gellardo** [comment]
No, these commands do not work. This is how it goes:

```
me@host ~/ % mkdir -p /tmp/tmp1/tmp2
me@host ~/ % z /tmp/tmp1/tmp2
me@host /tmp/tmp1/tmp2 % rm -rf /tmp/tmp1
me@host /tmp/tmp1/tmp2 % z ../..
zoxide: no match found
me@host /tmp/tmp1/tmp2 % z music
me@host ~/Music %
```

**Comment by ajeetdsouza** [comment]
Hm. that's not supposed to happen. It works fine for me on both Linux and macOS. Which OS are you on?

Could something be causing problems from your config? Try opening up a shell in debug mode with no default configuration:

```sh
zsh -x --no-rcs --no-globalrcs
```

Could you tell me the output of this now?

```sh
mkdir -p /tmp/tmp1/tmp2
z /tmp/tmp1/tmp2

rm -rf /tmp/tmp1
z ../.. # succeeds, takes me to /tmp
```

**Comment by Gellardo** [comment]
~Still does not work in the debug mode you mentioned.~
https://gist.github.com/Gellardo/89162aa12ffc8dc8dd361efcac8ecdb3

I think it works as intended:

By accident, I tried the same again in with only one addition: before going into `/tmp/tmp1/tmp2`, i first executed `z /tmp`, then continued as before.
Now everything worked as you are describing it.

I could still reproduce both cases (twice) - somehow after restarting the shell it "broke" again.
So my guess what is happening: I never visited /tmp directly since I started using zoxide.
That does not explain why i could reproduce the breaking case after the first successful one...

Anyways, thank you for your patience! I'm sufficiently convinced now that it was (unknowingly) a user error.

**Comment by ajeetdsouza** [comment]
`z ../..` should have worked even if the directory wasn't in your database. If `cd` works, then `z` should work as well.

I think I managed to isolate the issue from your logs. Apparently, this check appears to have failed for you:

> ```console
> $ mkdir -p /tmp/tmp1/tmp2
> $ cd /tmp/tmp1/tmp2
> $ rm -rf /tmp/tmp1
> $ [ -d ../.. ] && echo "success"
> ```

I tested this on macOS + zsh 5.8 and it prints out `success`, but it appears to be failing on your system somehow. I'm not sure why zsh thinks your `/tmp` is not a directory, but that's ideally what we should be investigating.
